WordPress/Classes/Services/MediaImageService.swift

```diff
@@ -0,0 +1,196 @@
+import Foundation
+
+/// A service for handling the process of retrieving and generating thumbnail images
+/// for existing Media objects, whether remote or locally available.
+final class MediaImageService: NSObject {
+    static let shared = MediaImageService()
+
+    private let coreDataStack: CoreDataStackSwift
+
+    init(coreDataStack: CoreDataStackSwift = ContextManager.shared) {
+        self.coreDataStack = coreDataStack
+    }
+
+    /// Returns a decompressed thumbnail for the given media asset.
+    ///
+    /// - Parameters:
+    ///   - media: The Media object.
+    ///   - preferredSize: An ideal size of the thumbnail in pixels. If `zero`, the maximum dimension of the UIScreen is used.
+    @MainActor
+    func image(for media: Media, preferredSize: CGSize) async throws -> UIImage {
+        guard media.remoteStatus != .stub else {
+            let media = try await fetchStubMedia(for: media)
+            // This should never happen, but adding it just in case to avoid recusion
+            guard media.remoteStatus != .stub else {
+                assertionFailure("The fetched media still has a .stub status")
+                throw MediaThumbnailExporter.ThumbnailExportError.failedToGenerateThumbnailFileURL
+            }
+            return try await image(for: media, preferredSize: preferredSize)
+        }
+
+        let imageURL = try await imageURL(for: media, preferredSize: preferredSize)
+        return try await Task.detached {
+            guard let image = UIImage(data: try Data(contentsOf: imageURL)) else {
+                throw URLError(.cannotDecodeContentData)
+            }
+            // Forces decompression (or bitmapping) to happen in the background.
+            // It's very expensive for some image formats, such as JPEG.
+            return image.preparingForDisplay() ?? image
```

Comment:
[Q] Is Task.detached {} what forces decompression to happen in the background?

Reply:
`Task.detached` sends the work to one of the background threads. It uses the new concurrency system, so you never risk running into a thread explosion problem, unlike `DispatchQueue.global`.

[`prepareForDisplay`](https://developer.apple.com/documentation/uikit/uiimage/3750834-preparingfordisplay) is a relatively new API (iOS 15) that decompressed the images.